src/DataService/DataObjectFactory.php

```diff
@@ -222,10 +219,9 @@ private function getBaseUri(DataService $annotation, IdInterface $account = null
         if (!($base = $annotation->getBaseUri())) {
             // If no account is specified, we must use the ResolveAllUrls service instead.
             if (!$account) {
-                /** @var ResolveAllUrls $urls */
-                $urls = ResolveAllUrls::load($this->authenticatedClient, $annotation->getService($readonly))->wait();
+                $resolver = new ResolveAllUrls($this->authenticatedClient, $this->cacheItemPool);
 
-                return $urls->resolve().$annotation->getPath();
+                return $resolver->resolve($annotation->getService($readonly))->resolve().$annotation->getPath();
```

Comment:
This line looks a little bit confusing. Semantically, it's strange that we use the resolve() method twice, it's like we are resolving it first, but not completely, and then we do it again, over a different returned object.

To me it's also confusing that we are using the word Resolve also in the classes, when it's a verb. The fact that you're using here and in the tests the variable $resolver, it's a clue that we should probably use something like that in the class name. Like I mentioned above, it's also strange that ResolveAllUrls::resolve() returns a ResolveAllUrlsResponse, that also have a resolve() method! And the ResolveAllUrlsResponse doesn't inherit ResolveBase, which can be confusing too. The fact that you're using $resolved, and $response as variable names to store that object, makes me thing that we should probably change the class name accordingly.

Oh! And it's curious that ResolveBase doesn't provide a resolve() method. But yeah, it's not an interface and there is no shared logic in the resolve for Domain and AllUrls right?

In summary, it looks to me that we're abusing the word resolve in methods and classes, and that we are using class names that sounds like verbs/actions, which is more appropriate for method names.

As a suggested solution, we could better use ResolverBase, DomainResolver, AllUrlsResolver, for example, or even a clearer set of words, if we find it. And I think we should change the ResolveAllUrlsResponse (which could be AllUrlsResolverResponse?) resolve method to a different, more specific, set of words. On the other hand, if resolve() has better meaning in the response class, maybe we could change and extend with other words the use of resolve in the other classes. In any case, I think we should avoid doing resolve()->resolve().

Reply:
> Semantically, it's strange that we use the resolve() method twice

Indeed. Does the rename to `getUrl` make this clearer?

> To me it's also confusing that we are using the word Resolve also in the classes, when it's a verb.

> As a suggested solution, we could better use ResolverBase, DomainResolver, AllUrlsResolver

My thinking was to match [the upstream mpx method names](https://docs.theplatform.com/help/wsf-resolveallurls-method). I agree, it would be better if they were REST-ish instead of RPC-ish in their naming, but I think it's better to match so anyone doing a class search based off of the mpx docs will find them.

> And the ResolveAllUrlsResponse doesn't inherit ResolveBase, which can be confusing too. 

Do think this it makes sense to move the response classes into their own directory / namespace?

> The fact that you're using $resolved, and $response as variable names to store that object, makes me thing that we should probably change the class name accordingly.

Oh, that's actually a bug in `ResolveDomain`. I mean, nothing is broken, but I didn't mean to overwrite the initial Guzzle response.

> Oh! And it's curious that ResolveBase doesn't provide a resolve() method. But yeah, it's not an interface and there is no shared logic in the resolve for Domain and AllUrls right?

Yeah, I thought about that, and the difference is that `resolveDomain` returns all options in one response, so we want to pass in the individual service we care about.

Reply:
You're right, matching should be a priority, otherwise it would be more confusing. The getUrl() renaming does make it clearer :+1: